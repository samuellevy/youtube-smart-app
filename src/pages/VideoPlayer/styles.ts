import styled from 'styled-components';
import { colors } from '../../global';

interface IOptionsBox {
  active?: boolean;
}

interface IButton {
  active?: boolean;
}

export const Container = styled.div`
  width: 100vw;
  height: 100vh;
  background-color: ${colors.grayBg};
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  iframe {
    width: 100vw;
    height: 100vh;
  }
`;

export const OptionsBox = styled.div<IOptionsBox>`
  width: 100vw;
  height: 100vh;
  background-color: ${colors.grayBg}aa;
  position: absolute;
  top: 0;
  left: 0;
  /* display: ${(props) => (props.active ? 'flex' : 'none')}; */
  display: flex;
  opacity: ${(props) => (props.active ? 1 : 0)};
  transition: opacity .5s ease;
  justify-content: center;
  align-items: center;
`;

export const Button = styled.button<IButton>`
  padding: 10px 15px;
  border-radius: 10px;
  background-color: ${(props) => (props.active ? colors.grayBg : colors.gray)};
  color: ${(props) => (props.active ? 'white' : 'black')};
  margin: 0 10px;
  font-family: 'Open Sans';
  font-size: 18px;
`;
