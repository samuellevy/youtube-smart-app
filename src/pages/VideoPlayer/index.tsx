import React, { createRef, useEffect, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useKeyboardContext } from '../../context/KeyboardContext';
import { addFavorite, useFavoritesContext } from '../../context/FavoriteContext';

import { Container, OptionsBox, Button } from './styles';
import { api } from '../../services/api';
import { IVideoEntityReq } from '../../dtos/IVideoEntityReq';

const VideoPlayer: React.FC = () => {
  const urlParams = useLocation().search;
  const [videoId, setVideoId] = useState('');
  const [loading, setLoading] = useState(true);
  const { keyboard }: any = useKeyboardContext();
  const { favorites, dispatch }: any = useFavoritesContext();
  const [optionsBoxActive, setOptionsBoxActive] = useState(false);
  const [activeButton, setActiveButton] = useState(0);
  const history = useHistory();
  const [data, setData] = useState<IVideoEntityReq>();

  useEffect(() => {
    controlHandler(keyboard.key);
  }, [keyboard]);

  useEffect(() => {
    setVideoId(urlParams.substr(1));
    setLoading(false);
    setOptionsBoxActive(false);

    getData(urlParams.substr(1));
  }, [urlParams]);

  useEffect(() => {
    console.log(favorites);
  }, [favorites]);

  const getData = async (id: string) => {
    const response = await api.get('/videos', {
      params: {
        id,
        part: 'snippet,statistics',
      },
    });

    setData(response.data.items[0]);
  };

  const controlHandler = (key: string) => {
    switch (key) {
      case 'Escape':
        if (!optionsBoxActive) {
          history.goBack();
        } else {
          setOptionsBoxActive(false);
        }
        break;
      case 'Enter':
        if (!optionsBoxActive) {
          setOptionsBoxActive(true);
        } else if (optionsBoxActive && activeButton === 0) {
          history.goBack();
        } else if (optionsBoxActive && activeButton === 1) {
          handleFavorite();
        }
        break;
      case 'ArrowLeft':
        if (optionsBoxActive) {
          setActiveButton(0);
        }
        break;
      case 'ArrowRight':
        if (optionsBoxActive) {
          setActiveButton(1);
        }
        break;
      default:
        break;
    }
  };

  const handleFavorite = () => {
    dispatch(addFavorite({
      id: videoId,
      channelTitle: data!.snippet.channelTitle,
      title: data!.snippet.title,
    }));
  };

  return (
    <Container>
      {
      loading ? <h1>Carregando...</h1>
        : <iframe title="yt-player" src={`https://www.youtube.com/embed/${videoId}?autoplay=1&controls=0`} frameBorder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowFullScreen />
      }
      <OptionsBox active={optionsBoxActive}>
        <Button active={activeButton === 0}>Continuar</Button>
        <Button active={activeButton === 1}>Adicionar aos favoritos</Button>
      </OptionsBox>
    </Container>
  );
};

export default VideoPlayer;
