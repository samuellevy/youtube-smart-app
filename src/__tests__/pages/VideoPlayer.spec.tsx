import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import VideoPlayer from '../../pages/VideoPlayer';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react-router-dom', () => ({
  useLocation: () => ({
    search: jest.fn,
  }),
  useHistory: () => ({
    push: jest.fn,
  }),
}));

describe('<VideoPlayer /> component', () => {
  it('should render properly', () => {
    const tree = renderer.create(<VideoPlayer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
