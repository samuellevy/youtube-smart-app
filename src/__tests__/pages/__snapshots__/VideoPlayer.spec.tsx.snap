// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoPlayer /> component should render properly 1`] = `
<div
  className="sc-bdfBwQ gHlEAy"
>
  <h1>
    Carregando...
  </h1>
  <div
    className="sc-gsTCUz kNSsXs"
  >
    <button
      className="sc-dlfnbm dUsTiH"
    >
      Continuar
    </button>
    <button
      className="sc-dlfnbm cCrvPm"
    >
      Adicionar aos favoritos
    </button>
  </div>
</div>
`;
